File: golang/tests/pkg/agentbay/context_sync_test.go
package agentbay_test

import (
	"encoding/json"
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
)

func TestNewUploadPolicy(t *testing.T) {
	policy := agentbay.NewUploadPolicy()

	assert.NotNil(t, policy)
	assert.True(t, policy.AutoUpload)
	assert.Equal(t, agentbay.UploadBeforeResourceRelease, policy.UploadStrategy)
}

func TestNewDownloadPolicy(t *testing.T) {
	policy := agentbay.NewDownloadPolicy()

	assert.NotNil(t, policy)
	assert.True(t, policy.AutoDownload)
	assert.Equal(t, agentbay.DownloadAsync, policy.DownloadStrategy)
}

func TestNewDeletePolicy(t *testing.T) {
	policy := agentbay.NewDeletePolicy()

	assert.NotNil(t, policy)
	assert.True(t, policy.SyncLocalFile)
}

func TestNewSyncPolicy(t *testing.T) {
	policy := agentbay.NewSyncPolicy()

	assert.NotNil(t, policy)
	assert.NotNil(t, policy.UploadPolicy)
	assert.NotNil(t, policy.DownloadPolicy)
	assert.NotNil(t, policy.DeletePolicy)
	assert.NotNil(t, policy.BWList)
	assert.NotNil(t, policy.BWList.WhiteLists)
	assert.Len(t, policy.BWList.WhiteLists, 1)
	assert.Equal(t, "", policy.BWList.WhiteLists[0].Path)
	assert.Empty(t, policy.BWList.WhiteLists[0].ExcludePaths)

}

func TestNewContextSync(t *testing.T) {
	contextID := "ctx-12345"
	path := "/home/user"
	policy := agentbay.NewSyncPolicy()

	sync, err := agentbay.NewContextSync(contextID, path, policy)

	assert.NoError(t, err)
	assert.NotNil(t, sync)
	assert.Equal(t, contextID, sync.ContextID)
	assert.Equal(t, path, sync.Path)
	assert.Equal(t, policy, sync.Policy)
}

func TestNewContextSyncWithNilPolicy(t *testing.T) {
	contextID := "ctx-12345"
	path := "/home/user"

	sync, err := agentbay.NewContextSync(contextID, path, nil)

	assert.NoError(t, err)
	assert.NotNil(t, sync)
	assert.Equal(t, contextID, sync.ContextID)
	assert.Equal(t, path, sync.Path)
	assert.Nil(t, sync.Policy)
}

func TestContextSyncWithPolicy(t *testing.T) {
	contextID := "ctx-12345"
	path := "/home/user"

	sync, err := agentbay.NewContextSync(contextID, path, nil)
	assert.NoError(t, err)
	assert.Nil(t, sync.Policy)

	// Create a custom policy
	policy := &agentbay.SyncPolicy{
		UploadPolicy: &agentbay.UploadPolicy{
			AutoUpload:     true,
			UploadStrategy: agentbay.UploadBeforeResourceRelease,
		},
		DownloadPolicy: &agentbay.DownloadPolicy{
			AutoDownload:     false,
			DownloadStrategy: agentbay.DownloadAsync,
		},
		DeletePolicy: &agentbay.DeletePolicy{
			SyncLocalFile: false,
		},
		BWList: &agentbay.BWList{
			WhiteLists: []*agentbay.WhiteList{
				{
					Path:         "/data",
					ExcludePaths: []string{"/data/temp"},
				},
			},
		},
	}

	// Apply the policy
	result, err := sync.WithPolicy(policy)

	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.Same(t, sync, result) // Verify method chaining returns the same instance
	assert.Equal(t, policy, sync.Policy)

	// Verify upload policy
	assert.NotNil(t, sync.Policy.UploadPolicy)
	assert.True(t, sync.Policy.UploadPolicy.AutoUpload)
	assert.Equal(t, agentbay.UploadBeforeResourceRelease, sync.Policy.UploadPolicy.UploadStrategy)

	// Verify download policy
	assert.NotNil(t, sync.Policy.DownloadPolicy)
	assert.False(t, sync.Policy.DownloadPolicy.AutoDownload)
	assert.Equal(t, agentbay.DownloadAsync, sync.Policy.DownloadPolicy.DownloadStrategy)

	// Verify delete policy
	assert.NotNil(t, sync.Policy.DeletePolicy)
	assert.False(t, sync.Policy.DeletePolicy.SyncLocalFile)

	// Verify BWList
	assert.NotNil(t, sync.Policy.BWList)
	assert.Len(t, sync.Policy.BWList.WhiteLists, 1)
	assert.Equal(t, "/data", sync.Policy.BWList.WhiteLists[0].Path)
	assert.Equal(t, []string{"/data/temp"}, sync.Policy.BWList.WhiteLists[0].ExcludePaths)

}

func TestDefaultSyncPolicyMatchesRequirements(t *testing.T) {
	// Create a default sync policy
	policy := agentbay.NewSyncPolicy()

	// Verify the policy matches the exact requirements
	assert.NotNil(t, policy)

	// Verify uploadPolicy
	assert.NotNil(t, policy.UploadPolicy)
	assert.True(t, policy.UploadPolicy.AutoUpload, "uploadPolicy.autoUpload should be true")
	assert.Equal(t, agentbay.UploadBeforeResourceRelease, policy.UploadPolicy.UploadStrategy, "uploadPolicy.uploadStrategy should be 'UploadBeforeResourceRelease'")

	// Verify downloadPolicy
	assert.NotNil(t, policy.DownloadPolicy)
	assert.True(t, policy.DownloadPolicy.AutoDownload, "downloadPolicy.autoDownload should be true")
	assert.Equal(t, agentbay.DownloadAsync, policy.DownloadPolicy.DownloadStrategy, "downloadPolicy.downloadStrategy should be 'DownloadAsync'")

	// Verify deletePolicy
	assert.NotNil(t, policy.DeletePolicy)
	assert.True(t, policy.DeletePolicy.SyncLocalFile, "deletePolicy.syncLocalFile should be true")

	// Verify bwList
	assert.NotNil(t, policy.BWList)
	assert.NotNil(t, policy.BWList.WhiteLists)
	assert.Len(t, policy.BWList.WhiteLists, 1, "bwList.whiteLists should have exactly 1 element")

	whiteList := policy.BWList.WhiteLists[0]
	assert.Equal(t, "", whiteList.Path, "bwList.whiteLists[0].path should be empty string")
	assert.NotNil(t, whiteList.ExcludePaths)
	assert.Empty(t, whiteList.ExcludePaths, "bwList.whiteLists[0].excludePaths should be empty array")

	// Additional verification: test JSON marshaling to ensure the structure is correct
	jsonData, err := json.Marshal(policy)
	assert.NoError(t, err, "Policy should be marshallable to JSON")

	// Debug: print the JSON structure
	t.Logf("Generated JSON: %s", string(jsonData))

	// Verify the JSON structure matches the expected format
	var jsonMap map[string]interface{}
	err = json.Unmarshal(jsonData, &jsonMap)
	assert.NoError(t, err, "JSON should be unmarshallable")

	// Verify uploadPolicy in JSON
	uploadPolicy, exists := jsonMap["uploadPolicy"].(map[string]interface{})
	assert.True(t, exists, "uploadPolicy should exist in JSON")
	assert.Equal(t, true, uploadPolicy["autoUpload"], "JSON uploadPolicy.autoUpload should be true")
	assert.Equal(t, "UploadBeforeResourceRelease", uploadPolicy["uploadStrategy"], "JSON uploadPolicy.uploadStrategy should be 'UploadBeforeResourceRelease'")

	// Verify downloadPolicy in JSON
	downloadPolicy, exists := jsonMap["downloadPolicy"].(map[string]interface{})
	assert.True(t, exists, "downloadPolicy should exist in JSON")
	assert.Equal(t, true, downloadPolicy["autoDownload"], "JSON downloadPolicy.autoDownload should be true")
	assert.Equal(t, "DownloadAsync", downloadPolicy["downloadStrategy"], "JSON downloadPolicy.downloadStrategy should be 'DownloadAsync'")

	// Verify deletePolicy in JSON
	deletePolicy, exists := jsonMap["deletePolicy"].(map[string]interface{})
	assert.True(t, exists, "deletePolicy should exist in JSON")
	assert.Equal(t, true, deletePolicy["syncLocalFile"], "JSON deletePolicy.syncLocalFile should be true")

	// Verify bwList in JSON
	bwList, exists := jsonMap["bwList"].(map[string]interface{})
	assert.True(t, exists, "bwList should exist in JSON")

	whiteLists, exists := bwList["whiteLists"].([]interface{})
	assert.True(t, exists, "bwList.whiteLists should exist in JSON")
	assert.Len(t, whiteLists, 1, "JSON bwList.whiteLists should have exactly 1 element")

	whiteListMap := whiteLists[0].(map[string]interface{})
	assert.Equal(t, "", whiteListMap["path"], "JSON bwList.whiteLists[0].path should be empty string")

	// Check if excludePaths exists in the JSON (it might be omitted if empty)
	if excludePaths, exists := whiteListMap["excludePaths"]; exists {
		excludePathsArray, ok := excludePaths.([]interface{})
		assert.True(t, ok, "bwList.whiteLists[0].excludePaths should be an array if present")
		assert.Empty(t, excludePathsArray, "JSON bwList.whiteLists[0].excludePaths should be empty array")
	} else {
		// If excludePaths is not present in JSON (omitted due to omitempty), that's also acceptable
		t.Log("excludePaths is omitted from JSON (which is acceptable due to omitempty tag)")
	}

	// Check that syncPaths should not exist in the JSON
	_, syncPathsExists := jsonMap["syncPaths"]
	assert.False(t, syncPathsExists, "syncPaths should not exist in JSON")
}

File: typescript/tests/unit/context-sync.test.ts
import {
  ContextSync,
  SyncPolicy,
  newSyncPolicy,
  newUploadPolicy,
  newDownloadPolicy,
  newDeletePolicy,
  newRecyclePolicy,
  newExtractPolicy,
  RecyclePolicy,
  UploadStrategy,
  DownloadStrategy,
  Lifecycle
} from "../../src/context-sync";
import { log } from "../../src/utils/logger";

describe("ContextSync Unit Tests", () => {
  describe("SyncPolicy Default Construction", () => {
    it("should create default SyncPolicy with correct values", () => {
      const policy = newSyncPolicy();


      // Verify the policy is not null
      expect(policy).toBeDefined();

      // Verify uploadPolicy
      expect(policy.uploadPolicy).toBeDefined();
      if (policy.uploadPolicy) {
        expect(policy.uploadPolicy.autoUpload).toBe(true);
        expect(policy.uploadPolicy.uploadStrategy).toBe(UploadStrategy.UploadBeforeResourceRelease);
      }

      // Verify downloadPolicy
      expect(policy.downloadPolicy).toBeDefined();
      if (policy.downloadPolicy) {
        expect(policy.downloadPolicy.autoDownload).toBe(true);
        expect(policy.downloadPolicy.downloadStrategy).toBe(DownloadStrategy.DownloadAsync);
      }

      // Verify deletePolicy
      expect(policy.deletePolicy).toBeDefined();
      if (policy.deletePolicy) {
        expect(policy.deletePolicy.syncLocalFile).toBe(true);
      }

      // Verify recyclePolicy
      expect(policy.recyclePolicy).toBeDefined();
      if (policy.recyclePolicy) {
        expect(policy.recyclePolicy.lifecycle).toBe(Lifecycle.Lifecycle_Forever);
        expect(policy.recyclePolicy.paths).toBeDefined();
        expect(policy.recyclePolicy.paths).toHaveLength(1);
        expect(policy.recyclePolicy.paths[0]).toBe("");
      }

      // Verify bwList
      expect(policy.bwList).toBeDefined();
      if (policy.bwList && policy.bwList.whiteLists) {
        expect(policy.bwList.whiteLists).toBeDefined();
        expect(policy.bwList.whiteLists).toHaveLength(1);
        
        const firstWhiteList = policy.bwList.whiteLists[0];
        expect(firstWhiteList.path).toBe("");
        expect(firstWhiteList.excludePaths).toBeDefined();
        if (firstWhiteList.excludePaths) {
          expect(firstWhiteList.excludePaths).toHaveLength(0);
        }
      }
      
    });

    it("should create default SyncPolicy that matches JSON requirements", () => {
      const policy = newSyncPolicy();

      // Convert to JSON and verify structure
      const jsonString = JSON.stringify(policy);
      const jsonObject = JSON.parse(jsonString);

      // Verify uploadPolicy in JSON
      expect(jsonObject.uploadPolicy).toBeDefined();
      expect(jsonObject.uploadPolicy.autoUpload).toBe(true);
      expect(jsonObject.uploadPolicy.uploadStrategy).toBe("UploadBeforeResourceRelease");

      // Verify downloadPolicy in JSON
      expect(jsonObject.downloadPolicy).toBeDefined();
      expect(jsonObject.downloadPolicy.autoDownload).toBe(true);
      expect(jsonObject.downloadPolicy.downloadStrategy).toBe("DownloadAsync");

      // Verify deletePolicy in JSON
      expect(jsonObject.deletePolicy).toBeDefined();
      expect(jsonObject.deletePolicy.syncLocalFile).toBe(true);

      // Verify recyclePolicy in JSON
      expect(jsonObject.recyclePolicy).toBeDefined();
      expect(jsonObject.recyclePolicy.lifecycle).toBe("Lifecycle_Forever");
      expect(jsonObject.recyclePolicy.paths).toBeDefined();
      expect(jsonObject.recyclePolicy.paths).toHaveLength(1);
      expect(jsonObject.recyclePolicy.paths[0]).toBe("");

      // Verify bwList in JSON
      expect(jsonObject.bwList).toBeDefined();
      expect(jsonObject.bwList.whiteLists).toBeDefined();
      expect(jsonObject.bwList.whiteLists).toHaveLength(1);
      expect(jsonObject.bwList.whiteLists[0].path).toBe("");
      // excludePaths should be present in JSON since it's an empty array
      expect(jsonObject.bwList.whiteLists[0].excludePaths).toBeDefined();
      expect(jsonObject.bwList.whiteLists[0].excludePaths).toHaveLength(0);

      // Verify syncPaths should not exist in JSON
      expect(jsonObject.syncPaths).toBeUndefined();

      // Log the generated JSON for verification
      log("Generated JSON:", jsonString);
    });

    it("should create individual policy components with correct defaults", () => {
      // Test UploadPolicy defaults
      const uploadPolicy = newUploadPolicy();
      expect(uploadPolicy.autoUpload).toBe(true);
      expect(uploadPolicy.uploadStrategy).toBe(UploadStrategy.UploadBeforeResourceRelease);

      // Test DownloadPolicy defaults
      const downloadPolicy = newDownloadPolicy();
      expect(downloadPolicy.autoDownload).toBe(true);
      expect(downloadPolicy.downloadStrategy).toBe(DownloadStrategy.DownloadAsync);

      // Test DeletePolicy defaults
      const deletePolicy = newDeletePolicy();
      expect(deletePolicy.syncLocalFile).toBe(true);
    });

    it("should create ContextSync with default policy", () => {
      const contextId = "test-context-123";
      const path = "/test/path";
      const policy = newSyncPolicy();

      const contextSync = new ContextSync(contextId, path, policy);

      expect(contextSync.contextId).toBe(contextId);
      expect(contextSync.path).toBe(path);
      expect(contextSync.policy).toBe(policy);
    });

    it("should allow chaining with withPolicy method", () => {
      const contextSync = new ContextSync("test-context", "/test/path");
      const policy = newSyncPolicy();

      const result = contextSync.withPolicy(policy);

      expect(result).toBe(contextSync);
      expect(contextSync.policy).toBe(policy);
    });

    it("should create RecyclePolicy with Lifecycle_1Day", () => {
      // Create a custom recycle policy with Lifecycle_1Day
      const customRecyclePolicy: RecyclePolicy = {
        lifecycle: Lifecycle.Lifecycle_1Day,
        paths: ["/custom/path"]
      };

      // Create a sync policy with the custom recycle policy
      const syncPolicy: SyncPolicy = {
        uploadPolicy: newUploadPolicy(),
        downloadPolicy: newDownloadPolicy(),
        deletePolicy: newDeletePolicy(),
        extractPolicy: newExtractPolicy(),
        recyclePolicy: customRecyclePolicy,
        bwList: {
          whiteLists: [
            {
              path: "",
              excludePaths: [],
            },
          ],
        },
      };

      // Verify the recycle policy
      expect(syncPolicy.recyclePolicy).toBeDefined();
      if (syncPolicy.recyclePolicy) {
        expect(syncPolicy.recyclePolicy.lifecycle).toBe(Lifecycle.Lifecycle_1Day);
        expect(syncPolicy.recyclePolicy.paths).toBeDefined();
        expect(syncPolicy.recyclePolicy.paths).toHaveLength(1);
        expect(syncPolicy.recyclePolicy.paths[0]).toBe("/custom/path");
      }

      // Test JSON serialization
      const jsonString = JSON.stringify(syncPolicy);
      const jsonObject = JSON.parse(jsonString);
      
      expect(jsonObject.recyclePolicy).toBeDefined();
      expect(jsonObject.recyclePolicy.lifecycle).toBe("Lifecycle_1Day");
      expect(jsonObject.recyclePolicy.paths).toHaveLength(1);
      expect(jsonObject.recyclePolicy.paths[0]).toBe("/custom/path");
      
    });

    it("should throw error when recyclePolicy path contains wildcard patterns", () => {
      // Create a recycle policy with invalid wildcard path
      const invalidRecyclePolicy: RecyclePolicy = {
        lifecycle: Lifecycle.Lifecycle_1Day,
        paths: ["/path/with/*"]
      };

      const syncPolicyWithInvalidPath: SyncPolicy = {
        uploadPolicy: newUploadPolicy(),
        downloadPolicy: newDownloadPolicy(),
        deletePolicy: newDeletePolicy(),
        extractPolicy: newExtractPolicy(),
        recyclePolicy: invalidRecyclePolicy,
        bwList: {
          whiteLists: [
            {
              path: "",
              excludePaths: [],
            },
          ],
        },
      };

      // Test that ContextSync constructor throws an error for invalid path
      expect(() => {
        new ContextSync("test-context", "/test/path", syncPolicyWithInvalidPath);
      }).toThrow("Wildcard patterns are not supported in path. Got: /path/with/*. Please use exact directory paths instead.");
    });
  });
});